### DISCUSSION TITLE
[Q&A] 회원가입 필요 정보와 구현 방법 관련 질문

### DISCUSSION THREAD

**Discussion by kang20** [toduck-App/toduck-backend#7] [Q&A]
### 자체 회원가입

![image](https://github.com/user-attachments/assets/5469809d-0761-4734-9ed6-c9d13afa118d)

- 이름 - 중복 X
- 휴대폰 번호 - 중복 X ( 휴대폰 인증까지 필수)
- 아이디 - 중복 X
- 비밀번호 - 중복허용, DB 저장시 인코딩, 필드값에
비밀번호와 비밀번호 확인 같은지 validation 백엔드에서 필요한지 생각

### Oauth
![image](https://github.com/user-attachments/assets/c22e4d60-a4a4-4c5f-aa48-a4de7ac95ba9)

→ Oauth 회원가입 후 추가 정보 필요한가?

현재 ERD 테이블 정보

![image](https://github.com/user-attachments/assets/7786a5f1-17ff-4a3d-9f31-eda8c30ea715)

## 2️⃣ 자체 회원가입 예상 로직

![image](https://github.com/user-attachments/assets/4e07dbaa-ff31-4b2a-9722-0f9ce0c1153b)

###  **AWS SMS 요금 부과 기준 (표준 주제 기준)**
![image](https://github.com/user-attachments/assets/82b7c3d9-ebe3-4d16-aa97-776243ccbae7)

### cool sms
![image](https://github.com/user-attachments/assets/19607662-9763-4011-831f-5906d90b3329)

### 네이버 클라우드 플랫폼 문자 발송
SMS, PUSH, 카카오 알림톡 등을 사용할 수 있는 서비스입니다. SMS의 경우 건당 9원, PUSH의 경우 건당 0.02원, 알림톡의 경우 건당 7.5원 등 비용이 발생합니다. (자세한 비용은 네이버 클라우드 플랫폼에서 확인 가능합니다.)

## ❗ 답변 포인트 
- 회원 가입시 필요한 정보에 대해 정리 해봤는데 이상한점이나 빠진 부분 위주로 봐주시면 됩니다.
- 자체 회원가입 구현 예상 로직 정리했는데 이상한점 확인해주세요
- 문자 서비스k 3개 정도 알아봤는데 어느 것을 선택할지 고민중입니다 리뷰를 보면 aws는 국내에서 쓰기 비싸다는 후기도 있고 cool sms 가 싸다는 후기도 있던데 문자 서비스에 대해 따로 아시는 정보나 위 3개 중에 어떤걸 선택할지 조언 부탁드립니다

**Comment by kang20** [comment]
+추가질문

1. 실명은 입력 받을 필요없지 않나?
2. 닉네임은 초기에 입력 무조건 안받는가? -> 받는다면 Oauth 도 입력 받아야함, 안받는다면 2다 임시 닉네임 자동으로 설정
3.  닉네임 제외하면 Oauth 는 추가 정보 필요 없는가?

**Comment by Seol-JY** [comment]
@kang20 

### 1.  
> 비밀번호와 비밀번호 확인 같은지 validation 백엔드에서 필요한지 생각

이부분은 별도로 필요 없다고 생각합니다. `Request`에서도 password 만 받으면 된다고 생각해요. 

> → Oauth 회원가입 후 추가 정보 필요한가?

회원 식별을 위해서 이메일을 따로 저장해야 하지 않을까라는 생각이 듭니다. 전화번호는 알 방법이 없어서.. 

### 2.
- 따로 적혀 있지는 않은데 전화번호, 아이디, 비밀번호 등 모든 사용자 입력에 대한 Validation 잘 고려해 주세요. (비밀번호 규칙 제한, 길이 제한 등)
- 전화번호 인증 및 중복 확인의 경우에서, 만료시간이 지났거나 존재하지 않는 경우에 대해서 통째로 실패응답을 반환하면 될 것 같아요.
- 마지막에 중복 여부 재 확인하는 과정은 좋은 것 같습니다. 회원가입 과정 중에 다른 사용자가 같은 이름으로 가입 한 경우가 있을 수 있으니까요.
- 전화번호 인증이 완료되었는지도 보안을 위해서는 최종 회원가입때 추가로 확인이 필요하다고 생각하는데요, 앞에서 전화번호 인증에 성공한 경우 일정 만료기간을 가진(30분 정도?) 일종의 토큰을 응답으로 반환하고, 최종 회원가입 때 같이 제출하도록 하는 방법이 있을 것 같아요. 전화번호를 인증 완료한 경우에 악의적인 사용자가 같은 전화번호로 인증 없이 가입을 시도할 수도 있으니까요.

### 3.
전화번호 인증에 대해서는 가격이 적절한 밴더사 찾으면 될 것 같습니다. 요청마다 요금이 발생하는 구조이니
1. 전화번호 당 하루 인증 가능 횟수를 제한 
2. 온전한 전화번호 형태가 아닌 경우(국내 기준)를 판별하기 위한 Validation 강화
등이 필요할 것 같습니다.

**Reply by kang20** [reply]
2,3 번은 확인 했습니다 그런데 만약 이메일을 따로 저장해야한다면 이메일 인증 과정 또한 필요할까요?

**Reply by Seol-JY** [reply]
@kang20 Ouath Provider 서버에서 이미 충분히 인증된 이메일을 제공 할 것이라고 생각합니다.

**Reply by kang20** [reply]
확인했습니다!

**Comment by Junad-Park** [comment]
@kang20 

### 1. 자체 회원가입
1. 이름 - 중복 x
    - 먼저 이름을 수집해야하는 이유를 모르겠습니다. 또한 동명이인을 고려하면 이름 중복을 허용하지 않는 것은 적절하지 않은 것 같습니다!
2. 휴대폰 전화번호 - 중복 x (휴대폰 인증까지 필수)
    - **동의합니다.**
3. 아이디 - 중복 x
    - **동의합니다.**
4. 비밀번호 - 중복허용, DB 저장시 인코딩, 필드값에 비밀번호와 비밀번호 확인 같은지 validation 백엔드에서 필요한지 생각
    - 프론트 측에서만 확인하면 된다고 생각합니다.

### 2. Oauth
1. Oauth 회원가입 후 추가 정보 필요한가?
    - 현재 모든 oauth provider 에게 공통으로 받을 수 있는 정보는 `1.email`, `2.이름/닉네임` 입니다. `이름/닉네임`은 중복이 발생할 수 있으므로 `email` 필드를 추가하는 것이 필요하다고 생각합니다. 

### 3. 문자 서비스
1. AWS SMS 서비스의 요금 문제
    - 물론 더 저렴하게 이용할 수 있는 곳이 있다면 교체하는 것도 하나의 방법이겠지만, @Seol-JY 님이 말씀하신 것 처럼 하루 횟수 제한등 정책을 활용해 해결하는 것도 괜찮다고 생각합니다.

**Reply by kang20** [reply]
확인했습니다!
